File: 2944-minimum-number-of-coins-for-fruits/2944-minimum-number-of-coins-for-fruits.java
class Solution {
    int[][] dp = new int[1001][1001];

    public int solve(int s, int[] v, int pow) {
        int n = v.length;
        if (s >= n)
            return 0;
        if (dp[s][pow] != -1)
            return dp[s][pow];
        if (pow == 0) {
            return v[s] + solve(s + 1, v, s + 1);
        }
        int ans = solve(s + pow, v, 0);
        for (int i = s; i <= Math.min(n - 1, s + pow - 1); i++) {
            ans = Math.min(ans, v[i] + solve(i + 1, v, i + 1));
        }
        return dp[s][pow] = ans;
    }
    public int minimumCoins(int[] prices) {
        int n = prices.length;
        for (int[] row : dp) {
            Arrays.fill(row, -1);
        }
        int ans = solve(0, prices, 0);
        return ans;
    }
}

File: 2944-minimum-number-of-coins-for-fruits/2944-minimum-number-of-coins-for-fruits.py
class Solution:
    def minimumCoins(self, prices: List[int]) -> int:
        queue, cost = collections.deque(), 0
        for index, price in enumerate(prices):
            while queue and queue[0][1] < index:
                queue.popleft()
            cost += price
            while queue and queue[-1][0] > cost:
                queue.pop()
            queue.append((cost, 2 * index + 1))
            cost = min(cost, queue[0][0])
        return cost